Bigtable: 01. Surface: Base structure

igorbernstein2 opened the pull request:
Layout the base structure to overlay the Bigtable GAPIC client. The general idea is for EnhancedBigtableStub to wrap the GAPIC GrpcBigtableStub
and for it to construct Callable chains on top of the generated chains.

This has been reviewed Kevin. However some things have changed since the original review:
- With the introduction of StubSettings all of the configuration has been moved to a new `EnhancedBigtableStubSettings`
- `BigtableDataSettings` has been updated to just delegate to `EnhancedBigtableStubSettings`
- `EnhancedBigtableStub` used to be called `BigtableDataStub`, but was renamed to avoid ambiguity

igorbernstein2 commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -0,0 +1,77 @@
+ */
+@InternalApi
+public class EnhancedBigtableStub implements AutoCloseable {
+  private final EnhancedBigtableStubSettings settings;
```
Please ignore, this PR is setting up structure that will be used in upcoming PRs

igorbernstein2 commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -0,0 +1,77 @@
+  private final EnhancedBigtableStubSettings settings;
+  private final GrpcBigtableStub stub;
+  private final ClientContext clientContext;
+  private final RequestContext requestContext;
```
Please ignore, this PR is setting up structure that will be used in upcoming PRs

igorbernstein2 commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -0,0 +1,77 @@
+public class EnhancedBigtableStub implements AutoCloseable {
+  private final EnhancedBigtableStubSettings settings;
+  private final GrpcBigtableStub stub;
+  private final ClientContext clientContext;
```
Please ignore, this PR is setting up structure that will be used in upcoming PRs

igorbernstein2 commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -0,0 +1,77 @@
+  private final EnhancedBigtableStubSettings settings;
+  private final GrpcBigtableStub stub;
+  private final ClientContext clientContext;
+  private final RequestContext requestContext;
```
Please ignore, this PR is setting up structure that will be used in upcoming PRs

igorbernstein2 commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -0,0 +1,77 @@
+public class EnhancedBigtableStub implements AutoCloseable {
+  private final EnhancedBigtableStubSettings settings;
+  private final GrpcBigtableStub stub;
+  private final ClientContext clientContext;
```
Please ignore, this PR is setting up structure that will be used in upcoming PRs

igorbernstein2 commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -0,0 +1,77 @@
+ */
+@InternalApi
+public class EnhancedBigtableStub implements AutoCloseable {
+  private final EnhancedBigtableStubSettings settings;
```
Please ignore, this PR is setting up structure that will be used in upcoming PRs

pongad commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -0,0 +1,175 @@
+     */
+    public Builder setInstanceName(InstanceName instanceName) {
+      Preconditions.checkNotNull(instanceName);
+      this.instanceName = instanceName;
```
You can `this.instanceName = Preconditions.checkNotNull(instanceName)`. Up to you though.

pongad commented on google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java:
```diff
@@ -0,0 +1,100 @@
+      error = t;
+    }
+
+    Truth.assertThat(error).isInstanceOf(IllegalStateException.class);
```
I think the recommendation is to `import static com.google.common.truth.Truth.assertThat` and just use `assertThat` in the tests.

igorbernstein2 commented:
All feedback has been incorporated. Also some minor changes applied:
- code reformatted via googleformat maven plugin
- Added Nonnull annotations to set{InstanceName,AppProfile}
- Suppressed unchecked warning
- Removed stale checked exception

Please merge when ready

igorbernstein2 commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -0,0 +1,77 @@
+  private final EnhancedBigtableStubSettings settings;
+  private final GrpcBigtableStub stub;
+  private final ClientContext clientContext;
+  private final RequestContext requestContext;
```
Please ignore, this PR is setting up structure that will be used in upcoming PRs

pongad commented:
LGTM. @garrettjonesgoogle PTAL

garrettjonesgoogle commented on google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -0,0 +1,175 @@
+ *   .build();
+ * }</pre>
+ */
+public class EnhancedBigtableStubSettings extends StubSettings<EnhancedBigtableStubSettings> {
```
Shouldn't this be `@InternalApi` like `EnhancedBigtableStub`?

igorbernstein2 commented:
Good catch! Updated, please merge
